docs/_includes/Statements.md

```diff
@@ -239,6 +239,153 @@ Note that the form
+```
+datatype VoidOutcome =
+| Success
+| Failure(error: string)
+{
+  predicate method IsFailure() {
+    this.Failure?
+  }
+  function method PropagateFailure(): VoidOutcome
+    requires IsFailure()
+  {
+  Failure(this.error); // this is Outcome<U>.Failure(...)
+  }
+}
+```
+Such a type also counts as a failure-compatible type for the purposes of a `:-` statement. However, now there is no `Extract` method and no value carried along with `Success`.
+The same exceptional return semantics apply. The only difference is that the list of LHS values now must omit what would be the first l-value, as there is no extracted value for it to get. Thus there may be no LHS values at all.
+
+In current Dafny, this form of exceptional return requires that the caller and callee each have only one output value, namely a value of this special failure-compatible extract-omitting type.
+
+TODO: Say more when the above restriction is lifted.
+
+### Chaining calls with potential failure
+
+Routinely using a first output value that has a failure-compatible type allows a programming style like that shown in the following example, using the `Outcome` class shown above.
+
+```
+method open(i: int) return (r: Outcome<int>) {
+  if (i < 0) return Failure("negative argument to a");
+  return Success(i+i);
+}
+
+method b(i: int) return (r: Outcome<bool>) {
+  if (i == 0) return Failure(zero argument to b");
+  return Success(100/i);
+}
+
+method c(i: int) return (r: Outcome<string>) {
+  if (i > 100) return Failure(big argument to c");
+  if (i > 0) Success("positive");
+  if (i < 0) Success("negative");
+  if (i == 0) Success("zero");
+}
+
+method m(i: int) return (r: Outcome<bool>){
+  var aa: int := a(i); // returns if i < 0
+  var bb: bool := b(i); // returns if i == 0
+  var cc: string := c(i); // returns if i > 100
+  return Success(true);
+}
+```
+
+TODO: Instead of requiring PropagateFailure, instead just desugar by calling
+CallerType.newFailure(this.error);
+
+
+
+
+
+### Failure returns and exceptions
+
+The `:-` mechanism for exceptional returns is like the exceptions used in other programming languages, with some similarities and differences.
+   * There is essentially just one kind of 'exception' in Dafny, the variations of the failure-compatible data type. Information about the failure is carried in the error message.
+   * A caller indicates that it accepts exceptional reeturns by declaring its first output to have a failure-compatible type. This is analogous to declaring that a method (e.g. in Java) might throw an exception.
+   * There is no such thing as an undeclared exception in Dafny. If a caller does not itself allow a failure output value, then that caller mayu nbot use `:-` and must receive any failure values from methods it calls using `:=`and then do something appropriate with the value. Because the failure value from the callee must be explicitly received, it cannot be silently ignored. (This is true of all the outputs from a called method.)
```

Comment:
```suggestion
   * There is no such thing as an undeclared exception in Dafny.
   If a caller does not itself allow a failure output value,
   then that caller may not use `:-` and must receive any failure values from methods it calls using `:=,
   and then do something appropriate with the value.
   Because the failure value from the callee must be explicitly received,
   it cannot be silently ignored.
   This is true of all the outputs from a called method.
```

Reply:
Revised/corrected/partially obsolete